alf/networks/encoding_networks.py

```diff
@@ -128,6 +128,81 @@ def forward(self, inputs, state=()):
+                 same_padding=False,
+                 activation=torch.relu_,
+                 kernel_initializer=None,
+                 flatten_output=False,
+                 name="ParallelImageEncodingNetwork"):
+        """
+        Args:
+            input_channels (int): number of channels in the input image
+            input_size (int or tuple): the input image size (height, width)
+            n (int): number of parallel networks
+            conv_layer_params (tuppe[tuple]): a non-empty tuple of
+                tuple (num_filters, kernel_size, strides, padding), where
+                padding is optional
+            same_padding (bool): similar to TF's conv2d ``same`` padding mode. If
+                True, the user provided paddings in `conv_layer_params` will be
+                replaced by automatically calculated ones; if False, it
+                corresponds to TF's ``valid`` padding mode (the user can still
+                provide custom paddings though)
+            activation (torch.nn.functional): activation for all the layers
+            kernel_initializer (Callable): initializer for all the layers.
+            flatten_output (bool): If False, the output will be an image
+                structure of shape ``BxnxCxHxW``; otherwise the output will be
+                flattened into a feature of shape ``BxnxN``.
+        """
+        input_size = _tuplify2d(input_size)
+        super().__init__(
+            input_tensor_spec=TensorSpec((input_channels, ) + input_size),
+            name=name)
+
+        assert isinstance(conv_layer_params, tuple)
+        assert len(conv_layer_params) > 0
+
+        self._flatten_output = flatten_output
+        self._conv_layer_params = conv_layer_params
+        self._conv_layers = nn.ModuleList()
+        for paras in conv_layer_params:
+            filters, kernel_size, strides = paras[:3]
+            padding = paras[3] if len(paras) > 3 else 0
+            if same_padding:  # overwrite paddings
+                kernel_size = _tuplify2d(kernel_size)
+                padding = ((kernel_size[0] - 1) // 2,
+                           (kernel_size[1] - 1) // 2)
+            self._conv_layers.append(
+                layers.ParallelConv2D(
+                    input_channels,
+                    filters,
+                    kernel_size,
+                    n,
+                    activation=activation,
+                    kernel_initializer=kernel_initializer,
+                    strides=strides,
+                    padding=padding))
+            input_channels = filters
+
+    def forward(self, inputs, state=()):
+        """The empty state just keeps the interface same with other networks."""
+        z = inputs
+        for conv_l in self._conv_layers:
+            z = conv_l(z)
+        if self._flatten_output:
+            z = torch.reshape(z, (*z.size()[0:2], -1))
+        return z, state
+
+
 @gin.configurable
 class ImageDecodingNetwork(Network):
```

Comment:
No parallel version for ImageDecodingNetwork?

Reply:
ParallelImageDecodingNetwork added